Add support for DependencyVersion to NuGet.exe cli UpdateCommand

ThomasSmeets opened the pull request:
Adding -DependencyVersion to NuGet Update

Linked to issue : https://github.com/NuGet/Home/issues/7694

Details: 
* Add new helper to handle the DependencyVersion, to avoid to duplicate code between the Install and Update command.
* Add the related unit tests to ensure the default behavior still works and the new features work as expect

## Testing/Validation

Tests Added: Yes
Reason for not adding tests:  
Validation:

dnfclas commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/NuGet/NuGet.Client?pullRequest=2800) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/NuGet/NuGet.Client?pullRequest=2800) before we can accept your contribution.<br/><br/>:x: ThomasSmeets [sign now](https://cla.dotnetfoundation.org/NuGet/NuGet.Client?pullRequest=2800)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/NuGet/NuGet.Client?pullRequest=2800) it.</sub>

nkolev92 commented on test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetUpdateCommandTests.cs:
```diff
@@ -1537,5 +1537,122 @@ public async Task UpdateCommand_Native_JS_Projects_Success()
+        {
+            // Test variations
+            // dependencyVersion, expected version
+            string[,] variations =
```
Use a Theory and providing the values there. 
Otherwise this provides a moving target if the tests ever fail. (you don't want a fail fast behavior here).

nkolev92 reviewed (commented):
All caveats of packages.config and nuget.exe update/install considered, this looks fine :) 
Please change the test to use a theory instead.

zivkan commented on src/NuGet.Clients/NuGet.CommandLine/Commands/InstallCommand.cs:
```diff
@@ -306,7 +278,7 @@ private DependencyBehavior GetDependencyBehavior()
             var allowPrerelease = Prerelease || (version != null && version.IsPrerelease);
 
-            var dependencyBehavior = GetDependencyBehavior();
+            var dependencyBehavior = DependencyBehaviorHelper.GetDependencyBehavior(DependencyBehavior.Lowest,DependencyVersion,Settings);
```
nitpick: formatting (whitespace after commas)

zivkan commented on src/NuGet.Clients/NuGet.CommandLine/Common/DependencyBehaviorHelper.cs:
```diff
@@ -0,0 +1,45 @@
+            return dependencyBehavior;
+        }
+
+        public static DependencyBehavior GetDependencyBehavior(DependencyBehavior defaultBehavior,string dependencyVersion, Configuration.ISettings settings)
```
nitpick: formatting (space after comma)

nkolev92 replied:
Format the whole document as it's a new one :)

zivkan commented on test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetUpdateCommandTests.cs:
```diff
@@ -1537,5 +1537,115 @@ public async Task UpdateCommand_Native_JS_Projects_Success()
+        [InlineData("Highest", "2.0.0")]
+        [InlineData("HighestMinor", "1.2.0")]
+        [InlineData("HighestPatch", "1.0.1")]
+        public async Task UpdateCommand_DependencyResolution_Success(string dependencyVersion, string expectedVersion)
```
nitpick: some parts of this method, particularly near the end, have consecutive empty lines.  It's not clear to me if there's a reason and no other tests in the file do it, so if there is a reason, it's better to use comments. Some tests have `// arrange`, `// act`, `// assert` comments to show the different phases of the test, in case that was the intention.

zivkan commented on test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetUpdateCommandTests.cs:
```diff
@@ -1537,5 +1537,115 @@ public async Task UpdateCommand_Native_JS_Projects_Success()
+                var packagesDirectory = Path.Combine(solutionDirectory, "packages");
+
+                // version installed will be the 1.1.0
+                var a1 = new PackageIdentity("A", new NuGetVersion("1.1.0"));
```
nitpick: I think `a1Identity` would be a better name, especially given there's a `a1Package` below, but I see that other tests in the file use the same name patterns, so I understand if you were copying them.

zivkan commented on test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetUpdateCommandTests.cs:
```diff
@@ -1537,5 +1537,115 @@ public async Task UpdateCommand_Native_JS_Projects_Success()
+                        new PackageDependencyGroup(nugetFramework,new List<Packaging.Core.PackageDependency>(){new Packaging.Core.PackageDependency("dep",new VersionRange(new NuGetVersion("1.0.0")))})
+                    });
+
+                var a2 = new PackageIdentity("A", new NuGetVersion("2.0.0"));
```
nitpick: similarly `a2Identity` would be more clear

zivkan commented on test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetUpdateCommandTests.cs:
```diff
@@ -1537,5 +1537,115 @@ public async Task UpdateCommand_Native_JS_Projects_Success()
+
+
+
+                var r = CommandRunner.Run(
```
nitpick: `result` might be a clearer name, but I know other tests just use `r`.

zivkan reviewed (approved):
First, thanks for the contribution. We appreciate you taking the time to help the product.

I didn't find any product issues, just some comments on formatting and naming (although the naming issues were copied from other tests, so it's great to keep the style of the existing code, it's just a shame the existing code could use improvement). I'll discuss with @nkolev92 if we want to wait for the formatting improvements or if we'll just merge it as is when we're ready.

ThomasSmeets commented:
Hello, 
I've changed the few formatting problems you have noticed.

nkolev92 commented on src/NuGet.Clients/NuGet.CommandLine/Common/DependencyBehaviorHelper.cs:
```diff
@@ -0,0 +1,47 @@
+
+            if (!string.IsNullOrEmpty(settingsDependencyVersion))
+            {
+                dependencyBehavior = TryGetDependencyBehavior(settingsDependencyVersion);
```
you can return here, and not even use the extra dependencyBehavior variable. 

Then at the bottom you just return defaultBehavior.

nkolev92 commented on src/NuGet.Clients/NuGet.CommandLine/NuGetResources.resx:
```diff
@@ -6182,4 +6178,8 @@ Oluşturma sırasında NuGet'in paketleri indirmesini önlemek için, Visual Stu
     <comment>Do not localize project.json and `PackageReference` 
 0 - project path (project.json file path)</comment>
   </data>
+  <data name="Error_UnknownDependencyVersion" xml:space="preserve">
```
I wouldn't rename the name of the resource. 
It'd break localization without more changes.

ThomasSmeets replied:
I can switch back to  **InstallCommandUnknownDependencyVersion** but, wouldn't it be strange to have a resource name called "Installxxx" even if this resource is used for "Update" behavior.

Furthermore I didn't found any localization for this resource (usually the resource seems localized with " ResourceName_Language"  but i didn't find any of them for  "InstallCommandUnknownDependencyVersion".
Do the localizations are contained in another project?

nkolev92 replied:
Nope, they are not, they usually in the same file. 
You're right this wouldn't affect localization so renaming it is ok. 
Nice catch :)

nkolev92 reviewed (commented):
A couple of small things.

nkolev92 commented:
fyi @karann-msft @aortiz-msft

zivkan commented on src/NuGet.Clients/NuGet.CommandLine/Common/DependencyBehaviorHelper.cs:
```diff
@@ -0,0 +1,44 @@
+using System;
```
Please add the same license header that you see in all other files

zivkan reviewed (commented):
I'm sorry we dropped the ball on this.

Can you please add the license header in the new files, rebase your changes onto the latest dev branch, and force push your branch? We can't merge it as it is now.

ThomasSmeets commented:
> I'm sorry we dropped the ball on this.
> 
> Can you please add the license header in the new files, rebase your changes onto the latest dev branch, and force push your branch? We can't merge it as it is now.

I am facing lot of problems to proceed to the rebase. ( mostly I am unable to rebase without requiring to resolve all the conflicts during the past year , which will certainly lead to errors). Any tips to make it easier ?

nkolev92 commented:
@ThomasSmeets 
Looking at your commit maybe you merged instead of rebased? I see the delta is huge here. 

If you can go to the `before rebase` state, then I'd recommend squashing all your commits into 1. 
Maybe even someone on the team can you do it for you if you can reset back to your before rebase state. We have push permissions to forks.

ThomasSmeets committed:
Add support for DependencyVersion to NuGet.exe cli UpdateCommand a363ddc17

- Add support for DependencyVersion  in UpdateCommand
-Add new helper to handle the DependencyVersion, to avoid to duplicate code between the Install and Update command.
-Add the related unit tests to ensure the default behavior still work and the new features work as expect

ThomasSmeets commented:
Had some problems with the rebase. 
But now it seems ok, you only have the a single commit with my wholes changes inside. I hope it will be fine.

zivkan commented:
Unfortunately our CI logs aren't public, so I need to copy details here. It's failing everywhere because of a compile error:

> ##[error]src\NuGet.Clients\NuGet.CommandLine\Common\DependencyBehaviorHelper.cs(20,27): Error CS0246: The type or namespace name 'CommandLineException' could not be found (are you missing a using directive or an assembly reference?)

> ##[error]src\NuGet.Clients\NuGet.CommandLine\Commands\UpdateCommand.cs(288,89): Error CS0103: The name 'DependencyBehavior' does not exist in the current context

ThomasSmeets committed:
Fix CI error

Add missing using  in UpdateCommand
Change change the CommandLineException which doesn't exist anymore in last version to CommandException in DependencyBehaviorHelper

zivkan commented:
@ThomasSmeets Thank you for your contribution. This should be available from NuGet 5.6 preview 2 and later.
